api/dao/containerutil.py

```
@@ -51,41 +51,43 @@
+        raise Exception('only one container can be specified')
     if query is None:
         query = {}
 
+    containers = ['acquisitions', 'sessions', 'subjects', 'projects', 'groups']
+    query.update({'parents.' + singularize(cont_name): cont_id})
+
     if include_refs:
         analysis_query = copy.deepcopy(query)
-        analysis_query.update({'parent.type': singularize(cont_name), 'parent.id': {'$in': cont_ids}})
+
         analysis_update = copy.deepcopy(update)
         analysis_update.get('$set', {}).pop('permissions', None)
         config.db.analyses.update_many(analysis_query, analysis_update)
 
         # Update job parents by destination
-        job_query = { 'parents.{}'.format(singularize(cont_name)) : {'$in': cont_ids} }
+        job_query = {'parents.{}'.format(singularize(cont_name)) :cont_id}
         config.db.jobs.update_many(job_query, analysis_update)
 
-    if cont_name in containers[:-1]:
-        child_cont = containers[containers.index(cont_name) + 1]
-        child_ids = [c['_id'] for c in config.db[child_cont].find({singularize(cont_name): {'$in': cont_ids}}, [])]
+    # Non standard containers only need to update related analysis and jobs, if any
+    if cont_name not in containers:
+        return
 
-        if child_ids:
-            child_query = copy.deepcopy(query)
-            child_query['_id'] = {'$in': child_ids}
-            config.db[child_cont].update_many(child_query, update)
-            # Recurse to the next hierarchy level
-            propagate_changes(child_cont, child_ids, query, update, include_refs=include_refs)
+    # TODO validate we dont send in invalid data in the update.  Can only be common data to the current level of hierarccy we are updating
+    for cur_cont in containers:
+        config.db[cur_cont].update_many(query, update)
+        if cont_name == cur_cont:
+            return
 
+    raise Exception('Never reached top level container from: {}'.format(cont_name))
```

The two "Developer Mistake" Exceptions in here add a bumps to the request for that specific type of Exception class. As is, they'll make it user-side if tripped.

---

Yes, they will make it user side if they fire.  They should never happen unless we code something incorrectly.  Which is your preference in this case?  I prefer to have it fail hard so that its very unlikely a coding error makes it into the wild.   I could also just log the error and use an exception type that is trapped in the handlers to prevent message from going to clients.  Whichever you prefer I can make that happen, or if you have an alternative approach I can do that too